### PR TITLE
allow for kafka-emitter to have extra dimensions be set for each event it emits

### PR DESCRIPTION
**Pull request by TSFenwick**
### Description

Allow for extra dimensions to be set in the kafka emitter. This functionality will allow the clusterName config for kafka-emitter to not need to necessarily be used as this new config allows for more flexibility to add n different dimensions.
Also made various nullable/nonnull annotation fixes

#### Release note

User configurable dimensions can be set via `druid.emitter.kafka.extra.dimensions` for events emitted by KafkaEmitter. For example, `druid.emitter.kafka.extra.dimensions={"region":"us-east-1","environment":"preProd"}`.

<hr>

##### Key changed/added classes in this PR
 * `KafkaEmitter`
 * `KafkaEmitterConfig`

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by TSFenwick** [comment]
running this locally to verify it works as expected

File: docs/development/extensions-contrib/kafka-emitter.md [hunk 1]
```diff
@@ -36,16 +36,17 @@ to monitor the status of your Druid cluster with this extension.
...
+| `druid.emitter.kafka.segmentMetadata.topic` | Kafka topic name for emitter's target to emit segment metadata. If `event.types` contains `segment_metadata`, this field cannot be empty.                  | no        | none                  |
+| `druid.emitter.kafka.producer.config`       | JSON configuration to set additional properties to Kafka producer.                                                                                         | no        | none                  |
+| `druid.emitter.kafka.clusterName`           | Optional value to specify the name of your Druid cluster. It can help make groups in your monitoring environment.                                          | no        | none                  |
+| `druid.emitter.kafka.extra.dimensions`      | Optional JSON configuration to specify a map of string extra dimensions for the events emitted. These can help make groups in your monitoring environment. | no        | none                  |
```
Comment (dkoepke) [inline docs/development/extensions-contrib/kafka-emitter.md:49]:
```suggestion
| `druid.emitter.kafka.extra.dimensions`      | Optional JSON configuration to specify a map of extra string dimensions for the events emitted. These can help make groups in your monitoring environment. | no        | none                  |
```

Comment (abhishekrb19) [inline docs/development/extensions-contrib/kafka-emitter.md:49]: If you reload your Intellij settings with https://github.com/apache/druid/blob/master/dev/druid_intellij_formatting.xml#L78, Intellij should stop auto-formatting the entire table.

File: extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitterConfig.java [hunk 2]
```diff
@@ -244,6 +256,7 @@ public String toString()
            ", request.topic='" + requestTopic + '\'' +
            ", segmentMetadata.topic='" + segmentMetadataTopic + '\'' +
            ", clusterName='" + clusterName + '\'' +
+           ", extra.dimensions='" + clusterName + '\'' +
```
Comment (abhishekrb19) [inline extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitterConfig.java:259]:
```suggestion
           ", extra.dimensions='" + extraDimensions + '\'' +
```

File: extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitterConfig.java [hunk 3]
```diff
@@ -62,7 +62,7 @@ public static EventType fromString(String name)
   public static final Set<EventType> DEFAULT_EVENT_TYPES = ImmutableSet.of(EventType.ALERTS, EventType.METRICS);
   @JsonProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG)
   private final String bootstrapServers;
-  @Nullable @JsonProperty("event.types")
+  @Nonnull @JsonProperty("event.types")
```
Comment (abhishekrb19) [inline extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitterConfig.java:65]: good catch 👍

File: extensions-contrib/kafka-emitter/src/test/java/org/apache/druid/emitter/kafka/KafkaEmitterConfigTest.java [hunk 4]
```diff
@@ -117,7 +120,7 @@ public void testSerDeNotRequiredKafkaProducerConfigOrKafkaSecretProducer()
   {
     KafkaEmitterConfig kafkaEmitterConfig = new KafkaEmitterConfig("localhost:9092", null, "metricTest",
                                                                    "alertTest", null, "metadataTest",
-                                                                   "clusterNameTest", null, null
+                                                                   "clusterNameTest", null, null, null
```
Comment (abhishekrb19) [inline extensions-contrib/kafka-emitter/src/test/java/org/apache/druid/emitter/kafka/KafkaEmitterConfigTest.java:123]: For code coverage, maybe make this non-null or add a new test with a non-null extra dimension map?

File: docs/development/extensions-contrib/kafka-emitter.md (hunk 1 above)
Comment (abhishekrb19) [inline docs/development/extensions-contrib/kafka-emitter.md:49]: The [Prometheus](https://druid.apache.org/docs/latest/development/extensions-contrib/prometheus/#configuration) emitter has a similar config `druid.emitter.prometheus.extraLabels`, so for consistency, call this `extraLabels` as well instead of `extra.dimensions`?



<br class="Apple-interchange-newline">

Comment (TSFenwick) [reply docs/development/extensions-contrib/kafka-emitter.md:49]: labels is a prometheus construct i chose dimensions because that is more of a druid construct.

Comment (TSFenwick) [reply docs/development/extensions-contrib/kafka-emitter.md:49]: [dogstatsd](https://github.com/apache/druid/blob/e184d24a74dbe425e2f80facbe09071049dcfb35/extensions-contrib/statsd-emitter/src/main/java/org/apache/druid/emitter/statsd/StatsDEmitterConfig.java#L70) has something similar as well named constant tags... granted its not a map, its just a list of tags

Comment (TSFenwick) [reply docs/development/extensions-contrib/kafka-emitter.md:49]: i have it configured. weird that its not using that :/

also i have turned that off a bunch manually but seems to do it randomly. Im guessing when i update intellij?

Comment (abhishekrb19) [reply docs/development/extensions-contrib/kafka-emitter.md]: Yeah, probably if you invalidate caches, etc? Alternatively, use vim, the old school way ;)

Comment (abhishekrb19) [reply docs/development/extensions-contrib/kafka-emitter.md]: Works for me. `extraDimensions`

File: docs/development/extensions-contrib/kafka-emitter.md [hunk 5]
```diff
@@ -36,16 +36,17 @@ to monitor the status of your Druid cluster with this extension.
...
+| Property                                    | Description                                                                                                                                                | Required | Default               |
+|---------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------|-----------------------|
+| `druid.emitter.kafka.bootstrap.servers`     | Comma-separated Kafka broker. (`[hostname:port],[hostname:port]...`)                                                                                       | yes       | none                  |
+| `druid.emitter.kafka.event.types`           | Comma-separated event types. <br/>Supported types are `alerts`, `metrics`, `requests`, and `segment_metadata`.                                             | no        | `["metrics", "alerts"]` |
```
Comment (zachjsh) [inline docs/development/extensions-contrib/kafka-emitter.md:42]: nit: Did the docs change for all of these properties? I find that when we update these tables, sometimes it updates all of the properties, if you didnt mean to update all the of the properties here would you mind reverting to makes doc review easier?

**Review by abhishekrb19 — APPROVED** [review]
LGTM. +1 after CI (checkstyle is failing), thanks @TSFenwick!

**Comment by abhishekrb19** [comment]
Ignoring the unrelated IT failure: `[standard-its / (Compile=openjdk8, Run=openjdk8, Cluster Build On K8s) ITNestedQueryPushDownTest integration test]` and the coverage check from hashCode, etc.
